packages/core/lib/phases.js

```diff
@@ -96,52 +99,62 @@ function createRamp(spec, ee) {
   const duration = spec.duration || 1;
   const arrivalRate = spec.arrivalRate;
   const rampTo = spec.rampTo;
+  const worker = spec.worker;
+  const totalWorkers = spec.totalWorkers;
 
-  // smallest tick we can get away with. Both arrivalRate and rampTo
-  // can be zero. So in that case we use 1s ticks even tho no
-  // arrivals will be generated
-  let tick = 1000 / Math.max(Math.max(arrivalRate, rampTo), 1);
   const difference = rampTo - arrivalRate;
-  const periods = duration * 1000 / tick;
-
-  function arrivalProbability(currentStep) {
-    // linear function ax + b
-    // normalized to 0 <= f(t) <= 1
-    // Anything under function value should be an arrival
+  const periods = duration;
+  debug(`worker ${worker} totalWorkers ${totalWorkers} arrivalRate ${arrivalRate} rampTo ${rampTo} difference ${difference} periods ${periods}`);
+
+  const periodArrivals = [];
+  const periodTick = [];
+  if (periods === 1) {
+    periodArrivals[0] = Math.floor((rampTo + arrivalRate) / 2);
+    periodTick[0] = 1000 / periodArrivals;
+  } else {
+    for (let i = 0; i < periods; i++) {
+      const rawPeriodArrivals = (difference / (duration - 1)) * i + arrivalRate;
+      periodArrivals[i] = Math.floor(rawPeriodArrivals);
+      if ((rawPeriodArrivals % 1) * totalWorkers * 1.1 >= worker) {
+        console.log(`worker ${worker} bumping period ${i} raw ${rawPeriodArrivals}`);
+        periodArrivals[i] = periodArrivals[i] + 1;
+      }
+      periodTick[i] = periodArrivals[i] > 0 ? Math.floor(1000 / periodArrivals[i]) : 1000;
+    }
+  }
 
-    let t = currentStep * tick / 1000;
-    return ((difference / duration) * t + arrivalRate) / Math.max(rampTo, arrivalRate) || 0;
-  };
+  debug(`periodArrivals ${periodArrivals}`);
+  debug(`periodTick ${periodTick}`);
 
-  let probabilities = Array.from({length: periods}, () => Math.random());
+  return async function rampTask(callback) {
+    let start = Date.now();
+    ee.emit('phaseStarted', spec);
+    for (let period = 0; period < periods; period++) {
+      ticker(period);
+      await sleep(1000);
+    }
 
-  debug(
-    `rampTo: tick = ${tick}; difference = ${difference}; rampTo = ${rampTo}; arrivalRate = ${arrivalRate}; periods = ${periods}`
-  );
+    const end = Date.now();
+    console.log(`Execution time: ${end - start} ms`);
```

Comment:
Execution time != phase time
Actually: execution time = duration+1 because driftless does not start instantly, it waits for the timer duration before executing the first time

Reply:
This was already happening prior to this (was actually a bigger difference)